E2e cv blog post part 3 (#3484)
iterative/dvc.org: content/blog/2022-05-09-end-to-end-computer-vision-api-part-3-remote-exp-ci-cd.md

---
title:
  'End-to-End Computer Vision API, Part 3: Remote Experiments & CI/CD For
  Machine Learning'
date: 2022-05-09
description: >
  In this final part, we will focus on leveraging cloud infrastructure with CML;
  enabling automatic reporting  (graphs, images, reports and tables with
  performance metrics) for PRs; and the eventual deployment process.
descriptionLong: |

>   In [Part 1](/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines),

Reviewer: you mean shortlinks? Does the frontmatter `descriptionLong` pick up link refs from the rest of the doc? i.e. this?

```suggestion
  In [Part 1][part 1],
```

>   In [Part 1](/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines),

Author: yeah, I misunderstood initially, but @shcheklein fixed it in https://github.com/iterative/dvc.org/pull/3529

---
title:
  'End-to-End Computer Vision API, Part 3: Remote Experiments & CI/CD For
  Machine Learning'
date: 2022-05-09
description: >
  In this final part, we will focus on leveraging cloud infrastructure with CML;
  enabling automatic reporting  (graphs, images, reports and tables with
  performance metrics) for PRs; and the eventual deployment process.
descriptionLong: |

>   In [Part 1](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines), 

Reviewer: ```suggestion
  In [Part 1](/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines),
```

  we talked about exploratory work in Jupyter Notebooks; versioning
  data in remote storage with DVC; and refactoring the code from Jupyter
  Notebooks into DVC pipeline stages.

>   [Part 2](https://dvc.org/blog/end-to-end-computer-vision-api-part-2-local-experiments) talked about the process of managing experiments with DVC pipelines,

Reviewer: ```suggestion
  [Part 2](/blog/end-to-end-computer-vision-api-part-2-local-experiments) talked about the process of managing experiments with DVC pipelines,
```

  DVCLive and Iterative Studio.

  In this final part, we will focus on leveraging cloud infrastructure with CML; 
  enabling automatic reporting 
  (graphs, images, reports and tables with performance metrics) for PRs; and
  the eventual deployment process.
picture: 2022-05-09/e2e-cv-pt3-cover.png
author: alex_kim
commentsUrl: https://discuss.dvc.org/t/end-to-end-computer-vision/1178
tags:
  - Computer Vision
  - DVC
  - CML
  - Studio
  - CI/CD
  - Experiment Tracking
---

### Leveraging Cloud Resources with CI/CD and CML

If you use the [CML library](https://cml.dev/) in combination with CI/CD tools
like GitHub Actions or GitLab CI/CD, you can quickly and easily:

1. provision a powerful virtual machine (VM) in the cloud as training Computer
   Vision (CV) models often requires powerful GPUs rarely available on local
   machines
2. submit your ML training job to it
3. save the results (metrics, models and other training artifacts)
4. automatically shut down the VM without having to worry about excessive cloud
   bills

> ![Continuous Integration and Deployment for Machine Learning](https://dvc.org/static/300c88b3b1b5f65753629d661cc916e5/2e49e/cicd4ml.png)

Reviewer: ```suggestion
![Continuous Integration and Deployment for Machine Learning](/img/cicd4ml.png '=460')
```

_Workflow for deploying to the production environment_

If there are no issues and we are confident in the quality of the new model, we
can merge the development branch into the main branch of our repository. Again,
this triggers the third CI/CD workflow that deploys the code from the main
branch to the production API. This looks identical to the deployment into the
development environment, except now the deployment endpoint will be

<https://demo-api-mag-tiles-prod.herokuapp.com/analyze> (note `-prod`).

## Summary

In this series of posts (see [Part 1] and [Part 2]), we described how we
addressed the problem of building a Computer Vision Web API for defect
detection. We’ve chosen this approach because it addresses the common challenges
that are shared across many CV projects: how to version datasets that consist of
a large number of small- to medium-sized files; how to avoid triggering
long-running stages of an ML pipeline when it’s not needed for reproducibility;
how to run model training jobs on the cloud infrastructure without having to
provision and manage everything yourself; and, finally, how to track progress in
key metrics when you run many ML experiments.

> [part 1]:
>   https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines
> [part 2]:
>   https://dvc.org/blog/end-to-end-computer-vision-api-part-2-local-experiments

Reviewer: ```suggestion
[part 1]: /blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines
[part 2]: /blog/end-to-end-computer-vision-api-part-2-local-experiments
```

We've talked about the following:

- Common difficulties when building Computer Vision Web API for defect detection

>   ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#introduction))
> - Pros and cons of exploratory work in Jupyter Notebooks
>   ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#proof-of-concept-in-jupyter-notebooks))
> - Versioning data in remote storage with DVC
>   ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#data-versioning))
> - Moving and refactoring the code from Jupyter Notebooks into DVC pipeline
>   stages
>   ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#refactoring-jupyter-code-into-an-ml-pipeline))
> - Experiment management and versioning
>   ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-2-local-experiments#experiment-management))
> - Visualization of experiments and collaboration in Iterative Studio
>   ([link](https://dvc.org/blog/end-to-end-computer-vision-api-part-2-local-experiments#collaboration-and-reporting-with-iterative-studio))

Reviewer: ```suggestion
  ([link](/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#introduction))
- Pros and cons of exploratory work in Jupyter Notebooks
  ([link](/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#proof-of-concept-in-jupyter-notebooks))
- Versioning data in remote storage with DVC
  ([link](/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#data-versioning))
- Moving and refactoring the code from Jupyter Notebooks into DVC pipeline
  stages
  ([link](/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines#refactoring-jupyter-code-into-an-ml-pipeline))
- Experiment management and versioning
  ([link](/blog/end-to-end-computer-vision-api-part-2-local-experiments#experiment-management))
- Visualization of experiments and collaboration in Iterative Studio
  ([link](/blog/end-to-end-computer-vision-api-part-2-local-experiments#collaboration-and-reporting-with-iterative-studio))
```

---
title:
  'End-to-End Computer Vision API, Part 3: Remote Experiments & CI/CD For
  Machine Learning'
date: 2022-05-09
description: >
  In this final part, we will focus on leveraging cloud infrastructure with CML;

>    enabling automatic reporting  (graphs, images, reports and tables with
>   performance metrics) for PRs; and the eventual deployment process.
> descriptionLong: |
> 
>   In [Part 1](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines), 
>   we talked about exploratory work in Jupyter Notebooks; versioning
>   data in remote storage with DVC; and refactoring the code from Jupyter
>   Notebooks into DVC pipeline stages.
> 
>   [Part 2](https://dvc.org/blog/end-to-end-computer-vision-api-part-2-local-experiments) talked about the process of managing experiments with DVC pipelines,
>   DVCLive and Iterative Studio.
> 
>   In this final part, we will focus on leveraging cloud infrastructure with CML; 
>   enabling automatic reporting 
>   (graphs, images, reports and tables with performance metrics) for PRs; and
>   the eventual deployment process.
> 

Reviewer: ```suggestion
  enabling automatic reporting  (graphs, images, reports and tables with
  performance metrics) for PRs; and the eventual deployment process.
descriptionLong: |
  In [Part 1](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines), 
  we talked about exploratory work in Jupyter Notebooks; versioning
  data in remote storage with DVC; and refactoring the code from Jupyter
  Notebooks into DVC pipeline stages.

  [Part 2](https://dvc.org/blog/end-to-end-computer-vision-api-part-2-local-experiments) talked about the process of managing experiments with DVC pipelines,
  DVCLive and Iterative Studio.

  In this final part, we will focus on leveraging cloud infrastructure with CML; 
  enabling automatic reporting 
  (graphs, images, reports and tables with performance metrics) for PRs; and
  the eventual deployment process.
```

_Output mask `exp3_num_258558_mask.png`_

Before we merge the dev branch into the main branch, we would need to thoroughly
test and monitor the application in the development environment. A good test
could be duplicating real image requests to the dev endpoint for some period of
time and assess the quality of the returned segmentation masks.

#### 3. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/3-deploy.yaml)

![Workflow for deploying to the production environment](/uploads/images/2022-05-09/workflow_prod.png '=400')
_Workflow for deploying to the production environment_

If there are no issues and we are confident in the quality of the new model, we
can merge the development branch into the main branch of our repository. Again,
this triggers the third CI/CD workflow that deploys the code from the main
branch to the production API. This looks identical to the deployment into the
development environment, except now the deployment endpoint will be

<https://demo-api-mag-tiles-prod.herokuapp.com/analyze> (note `-prod`).

## Summary

> In this series of posts (see
> [part 1](https://dvc.org/blog/end-to-end-computer-vision-api-part-1-data-versioning-and-ml-pipelines)
> and
> [part 2](https://dvc.org/blog/end-to-end-computer-vision-api-part-2-local-experiments)),

Reviewer: ```suggestion
In this series of posts (see [Part 1] and [Part 2]),
```
